### DISCUSSION TITLE
debugging complex boolean ops

### DISCUSSION THREAD

**Discussion by stewartoallen** [elalish/manifold#260] [General]
I've integrated Manifold into one of my more experimental projects, [a mesh editor](https://grid.space/mesh), which I often use as a playground for testing new capabilities that I want to add to my [slicer](https://grid.space/kiri). Using stock Manifold without my recent PR changes, I'm having difficulty debugging complex boolean on some objects. It's not clear to me how to proceed.

To start off, this is a simple test of union and subtraction.

<img width="1400" alt="Screen Shot 2022-10-24 at 3 54 13 PM" src="https://user-images.githubusercontent.com/554275/197616051-63d77fa2-d075-45f1-9d79-d2fc90288937.png">

This is a more complex test I have been doing with the jscad/modeling library involving 3DBenchy. Manifold just seems to ignore the benchy. But when I ask the API, it says all objects are manifold.

<img width="1400" alt="Screen Shot 2022-10-24 at 3 54 50 PM" src="https://user-images.githubusercontent.com/554275/197616324-5acf9855-2261-435c-82fe-2e7dd40ffe33.png">
<img width="1400" alt="Screen Shot 2022-10-24 at 3 55 32 PM" src="https://user-images.githubusercontent.com/554275/197616347-0271a499-06b4-4f24-859e-d2420f9b56b1.png">

I know it's not a memory or complexity issue, because Benchy only has 225K faces, and this next Twisted Flutes model has over 2M, and it completes both tests.

<img width="1400" alt="Screen Shot 2022-10-24 at 3 57 44 PM" src="https://user-images.githubusercontent.com/554275/197617706-172f9159-3e1a-4cf2-8179-d81daa1d7557.png">
<img width="1399" alt="Screen Shot 2022-10-24 at 3 59 07 PM" src="https://user-images.githubusercontent.com/554275/197617721-b00e4da1-63a9-4df8-88f1-b4865a28168c.png">
<img width="1399" alt="Screen Shot 2022-10-24 at 3 59 17 PM" src="https://user-images.githubusercontent.com/554275/197617733-63f1be17-c5f7-4b50-9e36-fb9a262714b0.png">

Any pointers appreciated!

**Comment by stewartoallen** [comment]
I should add that the mesh editor also has a repair feature that looks for, reports, and patches defects like open edges, dangling and invalid faces, etc. That code does not find any anomalies.

**Reply by elalish** [reply]
As far as meshes being "ignored", I'll just about guarantee it's that the input isn't manifold. In the event of non-manifold input, what you get is an empty mesh. What we're missing is a JS binding for the [`Status()`](https://manifoldcad.org/docs/html/classmanifold_1_1_manifold.html#a423e2f2a67ccc0662a97e018b2c3886d) method that tells you what check the input mesh failed. That should be easy to add. Just hadn't gotten to it since ManifoldCAD doesn't support import yet.

**Reply by stewartoallen** [reply]
it looks like `status()` is implemented and returns an int. in this case it's returning `2` for `non_manifold`. is it possible to get more detail such as the first face or vertex that or other feature that is flagged as non-manifold? that would be of interest to me so I can highlight it in the UI and give the user some ability to edit it.

**Reply by elalish** [reply]
Not currently; we early-out when finding a non-manifold edge, and an exhaustive list can be rather huge (especially for an STL). Open to ideas though.

**Reply by stewartoallen** [reply]
it's a little confusing because the `isManifold()` method returns `true` ... wasn't familiar enough with the api to look for `status()`

**Reply by stewartoallen** [reply]
I don't mind just getting the first thing to trip up the parser. It's at least a starting point for fixing something. I can also learn from that to improve my model sanity checking code.

**Reply by elalish** [reply]
Oh right, we gave you an empty mesh, which is manifold by default. Yeah, `IsManifold` isn't really meant to be a public method; it was for testing, but is probably not needed anymore. We should just remove it to avoid the confusion.

**Reply by elalish** [reply]
I would guess that debug will be easier if you write your own manifold checking function: there's not much to it. But if you come up with something that seems generically useful, go ahead a make a PR for it.

**Reply by stewartoallen** [reply]
it looks like normals inversion on some faces. this is something that's easy to highlight or fix by letting the user click one known good face and the engine can sort out the rest from there.

**Reply by elalish** [reply]
SG, as long as it's not a Klein bottle.
